draft-ietf-mls-protocol.md

```diff
@@ -1074,54 +1028,82 @@ the O(N) complexity of direct initialization. ]]
 
 ## Add
 
-An Add message is sent by a group member to add a new participant
-to the group.
+In order to add a new member to the group, an existing member of the
+group must take two actions:
+
+1. Send a Welcome message to the new member
+2. Send an Add message to the group (including the new member)
```

Comment:
The order is probably wrong, since the `Welcome` message should only be sent if the `Add` message succeeded. We should think some more about synchronicity issues that we want to avoid. 
Ideally the `Add` and `Welcome` message are on atomic operation in which all old and new members are notified at the same time. However since the existing member now waits for the `Add` to be successful before sending the (rather large) `Welcome`, there might be edge cases when the `Welcome` is never sent because the inviting member went offline too early.
For now I think we should keep this as an open issue. Going forward we should consider improvements, such as splitting the `Welcome` into a light header that gets reliably dispatched at the same time with the `Add` and a larger message containing the roster and the tree.

Reply:
Note that there's an open issue for this a few lines further down.  I agree that this is a tricky sync to get right.  Shoving the Welcome message in the Add message might simplify this, but the cost of sending an O(N) size message to everyone seems steep.